app/bundles/LeadBundle/Form/Type/CompanyType.php

```diff
@@ -0,0 +1,361 @@
+<?php
+/**
+ * @package     Mautic
+ * @copyright   2014 Mautic Contributors. All rights reserved.
+ * @author      Mautic
+ * @link        http://mautic.org
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+
+namespace Mautic\LeadBundle\Form\Type;
+
+use Mautic\CoreBundle\Factory\MauticFactory;
+use Mautic\CoreBundle\Form\EventListener\CleanFormSubscriber;
+use Mautic\CoreBundle\Form\EventListener\FormExitSubscriber;
+use Mautic\CoreBundle\Helper\DateTimeHelper;
+use Mautic\LeadBundle\Entity\Company;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+use Mautic\LeadBundle\Helper\FormFieldHelper;
+use Mautic\CoreBundle\Form\DataTransformer\IdToEntityModelTransformer;
+use Symfony\Component\Validator\Constraints\NotBlank;
+
+/**
+ * Class CompanyType
+ */
+class CompanyType extends AbstractType
+{
+    /**
+     * @var \Mautic\CoreBundle\Security\Permissions\CorePermissions
+     */
+    private $security;
+
+    /**
+     * @var \Symfony\Bundle\FrameworkBundle\Translation\Translator
+     */
+    private $translator;
+
+    private $factory;
+
+    /**
+     * @param MauticFactory $factory
+     */
+    public function __construct(MauticFactory $factory)
+    {
+        $this->translator = $factory->getTranslator();
+        $this->security = $factory->getSecurity();
+        $this->factory    = $factory;
+
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $fieldValues = (!empty($options['data'])) ? $options['data']->getFields() : ['filter' => ['isVisible' => true, 'object' => 'company']];
+        foreach ($options['fields'] as $field) {
```

Comment:
Let's push this to a trait method instead of duplicating the exact same code in LeadType and here.